src/mgr/MgrStandby.cc

```diff
@@ -161,6 +161,12 @@ int MgrStandby::init()
     client_messenger->wait();
     return r;
   }
+  /**
+   * this function must be called after mon.authenticate(); to avoid
+   * deadlock on MgrStandby::lock between MgrStandby::init() and
+   * MgrStandby::ms_dispatch()
+   */
```

Comment:
could you elaborate this dead lock a little bit. is it related this change?

at the first glance, i don't really think this call is necessary for fixing a deadlock. it is added to prevent monclient from swallowing the `MMonMap`, so that other dispatcher(s) subscribing message events can be notified on it. in this case, it's `Mgr`.

Reply:
At first, I thought monc.set_passthrough_monmap() is just setting a flag, so I added it next to monc.init(). The result is unexpected. I searched the code, but found nowhere is calling this function. With more debugs, I finally realize the reason and added this comment, hoping to help people avoid from the same trouble. But now it seems the comment brings more confuse than clarify. I prefer to remove this comment. What do you think?

Reply:
@cfanz i don't think you answered my query. 

> could you elaborate this dead lock a little bit. is it related this change?

probably you could put down the time sequence of the dead lock? it's important to us. because, if this dead lock is not related to this change, we might need to consider a fix, and will need to backport it.

Reply:
Oh, sorry. Here is how the dead lock happens if we call monc.set_passthrough_monmap() before monc.authenticate():
After monc.authenticate() is called, there will be several rounds messages come and go between monitor and monc, monc will wait on its own lock before the authentication is finished. However, monitor will send a monmap to monc before monc finally finish authenticate. Since passthrough flag is set, mgr will also handle this monmap message in  MgrStandby::ms_dispatch. Because mgr is still holding its lock in MgrStandby::init() due to the un-returned monc.authenticate(), MgrStandy::ms_dispatch has to wait for the lock, this wait in return will block the whole messenger's process thread. However monc still needs the messenger to continue process its unfinished authenticate stuff. 
Then deadlock happens.

Reply:
thanks! so it's only related to this change. that's a relief. i think you could rephrase it. like

> // only forward monmap updates after authentication finishes, otherwise
> // monc.authenticate() will be waiting for `MgrStandy::ms_dispatch()`
> // to acquire the lock forever, as it is already locked in the beginning of 
> // this method.